[BUG]: Fix ValueError in concat() when at least one Index has duplicates

phofl opened the pull request:
- [x] closes #36263
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

If obj_labes Index has duplicates and they are not removed from new_labels before redindexing, they are multiplied. So we would get a way too big index.

phofl committed:
Fix crash in concat if non unique index

jreback commented on pandas/core/reshape/concat.py:
```diff
@@ -496,7 +496,19 @@ def get_result(self):
                     obj_labels = obj.axes[1 - ax]
                     if not new_labels.equals(obj_labels):
-                        indexers[ax] = obj_labels.reindex(new_labels)[1]
+                        # We have to remove the duplicates from obj_labels
```
can't you do
`obj_labels = obj_labels.unique()` ? pandas keeps these in the same order

phofl replied:
No this does not work here unfortunately. 

The new_labels array may contain multiple duplicates. For example something like [1, 2, 2, 3, 3]
If obj_labels = [1, 2, 2, 3] we can only delete the duplicates which are also duplicated in obj_labels. The duplicates from the other inputs can not be deleted. That is the reason I did it this way.

phofl replied:
Sorry that did not answer your question. The reason we can not do ``obj_labels = obj_labels.unique()`` is different.

If new_labels = [1, 2, 2, 3, 3] and obj_labels = [1, 2, 2, 3]. if we would drop the duplicates, we would select the element at position 1 twice and the element at position 2 also twice, but for index 3. Also we would not select the element associated with index 3. We can not change the lenght of obj_labels without side effects, we have to modify new_labels

phofl replied:
Here is an example with obj_labels.unique()

```
    df1 = pd.DataFrame([1, 2, 3, 4], index=[0, 1, 1, 4], columns=["a"])
    df2 = pd.DataFrame([6, 7, 8, 9], index=[0, 0, 1, 3], columns=["b"])
    result = pd.concat([df1, df2], axis=1)
```
result in
```
    a    b
0  1.0  6.0
0  1.0  6.0
1  2.0  7.0
1  2.0  7.0
3  NaN  8.0
4  3.0  NaN
```

but clearly should equal
```
     a    b
0  1.0  6.0
0  1.0  7.0
1  2.0  8.0
1  3.0  8.0
3  NaN  9.0
4  4.0  NaN
```

jreback replied:
this is a highly sensitive part of the code so

a) make a function to do this. I *really* think you are duplicating a lot of existing code, but it could go in core/algorithms.py. I *know* we do this in places in core/indexing.py as well.
b) benchmark this.

phofl replied:
Could you give me a pointer as to where to look in indexing? I checked all functions in there but could not find a part where we do something similar like what we have to do here.

jreback commented:
also pls merge master

phofl committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into 36263

 Conflicts:
	doc/source/whatsnew/v1.2.0.rst
	pandas/tests/reshape/test_concat.py

phofl commented:
@jreback merged master

phofl committed:
Creta function in algos

phofl committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into 36263

phofl committed:
Add benchmark

phofl commented:
Moved it to algorithms and added a benchmark. I looked through indexing and the with index associated modules but could not find anything, which would help here. The check for uniqueness is not necessary from a technical standpoint, but it shoudl save some time for unique indices.

simonjayhawkins commented:
@phofl can you resolve merge conflicts

phofl committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into 36263

 Conflicts:
	doc/source/whatsnew/v1.2.0.rst
	pandas/core/algorithms.py
	pandas/tests/reshape/test_concat.py

phofl committed:
Add test in different file

phofl commented:
Done

phofl committed:
Fix pattern

jreback commented on pandas/core/reshape/concat.py:
```diff
@@ -501,6 +502,13 @@ def get_result(self):
+                        # duplicate or duplicates again
+                        if not obj_labels.is_unique:
+                            new_labels = algos.make_duplicates_of_left_unique_in_right(
+                                obj_labels.values, new_labels.values
```
use `np.asarray` instead of .values

phofl replied:
Thx, done

jreback commented on pandas/core/algorithms.py:
```diff
@@ -2149,3 +2149,21 @@ def _sort_tuples(values: np.ndarray[tuple]):
     return values[indexer]
+
+
+def make_duplicates_of_left_unique_in_right(left, right) -> np.ndarray:
```
can you type the input args.

can you add an example here.

is this unit tested?

phofl replied:
Added unittests, forgot them obviously, and typed the inputs. Example is below

jreback commented on pandas/tests/reshape/concat/test_concat.py:
```diff
@@ -556,3 +556,15 @@ def test_concat_preserves_extension_int64_dtype():
+        {"a": [1, 1, 2, 3, np.nan, 4], "b": [6, 7, 8, 8, 9, np.nan]},
+        index=Index([0, 0, 1, 1, 3, 4]),
+    )
+    tm.assert_frame_equal(result, expected)
```
extra file added below

phofl replied:
Moved the test if that is what you are referring to

phofl committed:
Use asarray

phofl committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into 36263

phofl commented:
Example:

We have the 2 DataFrames
```
df1 = DataFrame([1, 2, 3, 4], index=[0, 1, 1, 4], columns=["a"])
df2 = DataFrame([6, 7, 8, 9], index=[0, 0, 1, 3], columns=["b"])
```

When looping over ``df1`` (df2 case is similar)
In this case ``new_labels`` before calling ``make_duplicates_of_left_unique_in_right`` will be
```
new_labels=Index([0, 0, 1, 1, 3, 4])
```

while ``obj_labels`` will be
```
obj_labels=Index([0, 1, 1, 4])
```

Simply reindexing would return an indexer which would duplicate the pair ``[1, 1]`` again: ``np.array([0, 0, 1, 2, 1, 2, -1, 3])``. (taking every ``1`` twice)
The new function takes the duplications of the 1 in ``new_labels`` away.

``make_duplicates_of_left_unique_in_right`` returns 
```
new_labels=Index([0, 0, 1, 3, 4])
```

which leads us to the desired reindexing result ``np.array([0, 0, 1, 2, -1, 3])``

phofl committed:
Adress review comments

phofl committed:
Delete brackets

jreback reviewed (changes requested):
pandas/tests/reshape/test_concat.py  seems to be added here (i think we moved this), can you remove; pls merge master and ping on green.

cc @jbrockmendel if comments.

jbrockmendel commented on doc/source/whatsnew/v1.2.0.rst:
```diff
@@ -567,6 +567,7 @@ Reshaping
 - Bug in :meth:`DataFrame.combine_first()` caused wrong alignment with dtype ``string`` and one level of ``MultiIndex`` containing only ``NA`` (:issue:`37591`)
 - Fixed regression in :func:`merge` on merging DatetimeIndex with empty DataFrame (:issue:`36895`)
 - Bug in :meth:`DataFrame.apply` not setting index of return value when ``func`` return type is ``dict`` (:issue:`37544`)
+- Bug in :func:`concat` resulted in a ``ValueError`` when at least one of both inputs had a non unique index (:issue:`36263`)
```
non-unique

phofl replied:
Done

jbrockmendel commented on pandas/core/algorithms.py:
```diff
@@ -2149,3 +2149,23 @@ def _sort_tuples(values: np.ndarray[tuple]):
+    Duplicates of left are unique in right
+    """
+    left_duplicates = unique(left[duplicated(left)])
+    return right[~(duplicated(right) & np.isin(right, left_duplicates))]
```
any reason to prefer np.isin vs the algos.isin?

phofl replied:
Not that i remember, changed it.

jbrockmendel commented on pandas/core/algorithms.py:
```diff
@@ -2149,3 +2149,23 @@ def _sort_tuples(values: np.ndarray[tuple]):
+    Parameters
+    ----------
+    left: ndarray
+    right: ndarray
```
dtypes unrestricted?

phofl replied:
Could not think of anything, why they should be restricted.

jbrockmendel commented on pandas/tests/reshape/concat/test_dataframe.py:
```diff
@@ -167,3 +167,14 @@ def test_concat_dataframe_keys_bug(self, sort):
+        result = concat([df1, df2], axis=1)
+        expected = DataFrame(
+            {"a": [1, 1, 2, 3, np.nan, 4], "b": [6, 7, 8, 8, 9, np.nan]},
+            index=Index([0, 0, 1, 1, 3, 4]),
```
just to make sure i understand  this, for any `df1` and `df2`, we want `result.index` to always satisfy:

```
vc = result.index.value_counts()
vc1 = df1.index.value_counts()
vc2 = df2.index.value_counts()

vc1b = vc1.reindex(vc.index, fill_value=0)
vc2b = vc2.reindex(vc.index, fill_value=0)
```

We expect `vc` to be the pointwise maximum of `vc1b` and `vc2b`?

phofl replied:
Yes exactly. Thats perfectly on point.

jbrockmendel commented on pandas/core/algorithms.py:
```diff
@@ -2149,3 +2149,23 @@ def _sort_tuples(values: np.ndarray[tuple]):
     return values[indexer]
+
+
+def make_duplicates_of_left_unique_in_right(
```
is this related to or useful for the index.union-with-duplicates stuff?

phofl replied:
If you pass in the union as left and right, you would get the distinct result. Have to take a look if we can use this.

jbrockmendel commented on pandas/core/algorithms.py:
```diff
@@ -2149,3 +2149,23 @@ def _sort_tuples(values: np.ndarray[tuple]):
+) -> np.ndarray:
+    """
+    Drops all duplicates values from left in right, so that they are
+    unique in right.
```
The code itself looks good, but this sentence isn't clear to a reader without context

phofl replied:
Improved it?

phofl committed:
Change whatsnew

phofl committed:
Remove file

phofl committed:
Add comment

phofl committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into 36263

phofl commented:
Removed the concat file, probably got in through merging master the last time

jreback commented:
thanks @phofl very nice

phofl commented:
Thx, will try to use this for the Index.union with duplicates problems

jreback commented:
reverted here: https://github.com/pandas-dev/pandas/pull/38654
